orchagent/vxlan_tunnel_decap_orch.cpp

```diff
@@ -0,0 +1,215 @@
+                                tunnel_map_attrs.data()
+                          );
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        throw std::runtime_error("Can't create tunnel map object");
+    }
+
+    return tunnel_map_id;
+}
+
+static sai_object_id_t
+create_decap_tunnel_map_entry(
+    sai_object_id_t tunnel_map_id,
+    sai_uint32_t vni,
+    sai_uint16_t vlan_id)
+{
+    sai_attribute_t attr;
+    std::vector<sai_attribute_t> tunnel_map_entry_attrs;
+
+    attr.id = SAI_TUNNEL_MAP_ENTRY_ATTR_TUNNEL_MAP_TYPE;
+    attr.value.s32 = SAI_TUNNEL_MAP_TYPE_VNI_TO_VLAN_ID;
+    tunnel_map_entry_attrs.push_back(attr);
+
+    attr.id = SAI_TUNNEL_MAP_ENTRY_ATTR_TUNNEL_MAP;
+    attr.value.oid = tunnel_map_id;
+    tunnel_map_entry_attrs.push_back(attr);
+
+    attr.id = SAI_TUNNEL_MAP_ENTRY_ATTR_VNI_ID_KEY;
+    attr.value.u32 = vni;
+    tunnel_map_entry_attrs.push_back(attr);
+
+    attr.id = SAI_TUNNEL_MAP_ENTRY_ATTR_VLAN_ID_VALUE;
+    attr.value.u16 = vlan_id;
+    tunnel_map_entry_attrs.push_back(attr);
+
+    sai_object_id_t tunnel_map_entry_id;
+    sai_status_t status = sai_tunnel_api->create_tunnel_map_entry(
+                                &tunnel_map_entry_id,
+                                gSwitchId,
+                                static_cast<uint32_t>(tunnel_map_entry_attrs.size()),
+                                tunnel_map_entry_attrs.data()
+                          );
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        throw std::runtime_error("Can't create a tunnel map entry object");
+    }
+
+    return tunnel_map_entry_id;
+}
+
+// Create Tunnel
+static sai_object_id_t
+create_tunnel(sai_object_id_t tunnel_decap_map_id)
+{
+    sai_attribute_t attr;
+    std::vector<sai_attribute_t> tunnel_attrs;
+
+    attr.id = SAI_TUNNEL_ATTR_TYPE;
+    attr.value.s32 = SAI_TUNNEL_TYPE_VXLAN;
+    tunnel_attrs.push_back(attr);
+
+    sai_object_id_t decap_list[] = { tunnel_decap_map_id };
+    attr.id = SAI_TUNNEL_ATTR_DECAP_MAPPERS;
+    attr.value.objlist.count = 1;
+    attr.value.objlist.list = decap_list;
+    tunnel_attrs.push_back(attr);
+
+    sai_object_id_t tunnel_id;
+    sai_status_t status = sai_tunnel_api->create_tunnel(
+                                &tunnel_id,
+                                gSwitchId,
+                                static_cast<uint32_t>(tunnel_attrs.size()),
+                                tunnel_attrs.data()
+                          );
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        throw std::runtime_error("Can't create a tunnel object");
+    }
+
+    return tunnel_id;
+}
+
+// Create tunnel termination
+
+static sai_object_id_t
+create_tunnel_termination(
+    sai_object_id_t tunnel_oid,
+    sai_ip4_t dstip,
+    sai_object_id_t default_vrid)
+{
+    sai_attribute_t attr;
+    std::vector<sai_attribute_t> tunnel_attrs;
+
+    attr.id = SAI_TUNNEL_TERM_TABLE_ENTRY_ATTR_TYPE;
+    attr.value.oid = SAI_TUNNEL_TERM_TABLE_ENTRY_TYPE_P2MP;
```

Comment:
Should this be:
'attr.value.s32 = SAI_TUNNEL_TERM_TABLE_ENTRY_TYPE_P2MP;'?

Reply:
Yes. It should. You're right. I've fixed that.
Thank you for letting me know